pulsar-client-cpp/lib/ProducerImpl.cc

```diff
@@ -204,7 +206,8 @@ void ProducerImpl::handleCreateProducer(const ClientConnectionPtr& cnx, Result r
         }
 
         // if the producer is lazy the send timeout timer is already running
-        if (!conf_.getLazyStartPartitionedProducers()) {
+        if (!(conf_.getLazyStartPartitionedProducers() &&
+              conf_.getAccessMode() == ProducerConfiguration::Shared)) {
```

Comment:
Because the previous operation check the access mode.
```
if (conf_.getLazyStartPartitionedProducers() && conf_.getAccessMode() == ProducerConfiguration::Shared) {
        // we need to kick it off now as it is possible that the connection may take
        // longer than sendTimeout to connect
        startSendTimeoutTimer();
```

Reply:
I got it. I think it's because the previous implementation of lazy producer is somehow incorrect.

In Java client, when a `ProducerImpl` is created, it starts a send timer and call `grabCnx()`. So for a `PartitionedProducerImpl` object, Java client doesn't create a `ProducerImpl` if lazy start is enabled and the access mode is `Shared`.

However, in C++ client, the send timer is started after `grabCnx()` is called, which are both called in `ProducerImpl::start`, while the creation of a `ProducerImpl` doesn't do anything.

Regarding this PR, I think you can just keep the existing logic.

Reply:
So in C++ client, it checks `conf_.getLazyStartPartitionedProducers()` to verify if the owner of `ProducerImpl` is a `PartitionedProducerImpl`. This check is wrong. Because for a producer on a non-partitioned topic, `conf_.getLazyStartPartitionedProducers()` should not affect the existing logic.

There should be a method to check if a `ProducerImpl` is created from a `PartitionedProducerImpl`. I think we can check the `partition_` field.

Anyway, it doesn't affect this PR.